examples/pennylane_run_doubly_stochastic.py

```diff
@@ -0,0 +1,312 @@
+# to a reasonably close approximation of ground-state energy!
+
+##############################################################################
+# Doubly stochastic gradient descent for VQE
+# ------------------------------------------
+#
+# As noted in `Sweke et al. (2019) <https://arxiv.org/abs/1910.01155>`__,
+# variational quantum algorithms often include terms consisting of linear combinations
+# of expectation values. This is true of the parameter-shift rule (where the
+# gradient of each parameter is determined by shifting the parameter by macroscopic
+# amounts and taking the difference), as well as VQE, where the Hamiltonian
+# is usually decomposed into a sum of Pauli expectation values.
+#
+# Consider the Hamiltonian from the previous section. As this Hamiltonian is an
+# Hermitian observable, we can always express it as a sum of Pauli matrices using
+# the relation
+#
+# .. math::
+#
+#     H = \sum_{i,j=I,x,y,z} a_{i,j} (\sigma_i\otimes \sigma_j),
+#         ~~a_{i,j} = \frac{1}{4}\text{tr}[(\sigma_i\otimes \sigma_j )H].
+#
+# Applying this, we can see that
+#
+# .. math::
+#
+#     H = 4  + 2I\otimes X + 4I \otimes Z - X\otimes X + 5 Y\otimes Y + 2Z\otimes X.
+#
+# To perform "doubly stochastic" gradient descent, we simply apply the stochastic
+# gradient descent approach from above, but in addition also uniformly sample
+# a subset of the terms for the Hamiltonian expectation at each optimization step.
+# This inserts another element of stochasticity into the system---all the while
+# convergence continues to be guaranteed!
+#
+# Let's create a QNode that randomly samples three terms from the above
+# Hamiltonian as the observable to be measured.
+
+I = np.identity(2)
+X = np.array([[0, 1], [1, 0]])
+Y = np.array([[0, -1j], [1j, 0]])
+Z = np.array([[1, 0], [0, -1]])
+
+terms = np.array(
+    [2 * np.kron(I, X), 4 * np.kron(I, Z), -np.kron(X, X), 5 * np.kron(Y, Y), 2 * np.kron(Z, X)]
+)
+
+
+@qml.qnode(dev_SGD)
+def sampled_terms(params, A=None):
+    StronglyEntanglingLayers(*params, wires=[0, 1])
+    return expval(qml.Hermitian(A, wires=[0, 1]))
+
+
+def loss(params):
+    n = np.random.choice(np.arange(5), size=3, replace=False)
+    A = np.sum(terms[n], axis=0)
+    return 4 + (5 / 3) * sampled_terms(params, A=A)
+
+
+##############################################################################
+# Optimizing the circuit using gradient descent via the parameter-shift rule:
+
+dev_SGD.shots = 100
+cost = []
+params = init_params
+opt = qml.GradientDescentOptimizer(0.005)
+
+for _ in range(250):
+    cost.append(loss(params))
+    params = opt.step(loss, params)
+
+##############################################################################
+# Plotting the cost versus optimization step:
+
+
+def moving_average(data, n=3):
+    ret = np.cumsum(data, dtype=np.float64)
+    ret[n:] = ret[n:] - ret[:-n]
+    return ret[n - 1 :] / n
+
+
+average = np.vstack([np.arange(25, 200), moving_average(cost, n=50)[:-26]])

```

Comment:
I would either compute the analytic cost at each step or have some text that explains why the moving average is necessary and why the points in the plot go below the ground state energy. I actually prefer the latter here because it fits better to realistic application scenarios where you don't want to go analytic at each step.

Reply:
Done!